**In this assignment, you'll get to practice the concepts and skills covered in the course so far. The main objective of this assignment is to implement and use some of the tools, algorithms, and techniques to perform sentiment analysis on textual data..**



**Guidelines**
* Download `AmazonReviews.csv` file from D2L. 
* Make sure to run all the code cells, otherwise you may get errors like `NameError` for undefined variables.
* Do not change variable names, delete cells or disturb other existing code. It may cause problems during evaluation.
* In some cases, you may need to add some code cells or new statements before or after the line of code containing the `???`.
* Use markdown cells to write your discussions and reflections. 

**Procedure**
* Save your work as `IPYNB` file named `Lab8.ipynb` and submit to D2L `Lab 8 - Sequence Modeling (Dropbox)` by the due date.
* As you go through this notebook, you will find the symbol `???` in certain places. To complete this assignment, you must replace all the `???` with appropriate values, expressions or statements to ensure that the notebook runs properly end-to-end.
* Include your response for `Part 1` and `Part 2` in this notebook. 

<div class="alert alert-block alert-info">

# Part 1: Activity 

</div>

#### Read the content of the `AmazonReviews.csv` into a dataframe called `reviews_df` and perform the following:
1. **Preprocess the `reviews_df` dataframe to prepare it for the following questions.**

2. **Build a SimpleRNN Network to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy** 

3. **Build a Gated Recurrent Unit to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy** 

4. **Build a Long Short-Term Memory Model to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy** 

5. **Which model performs better and why?** 

In [ ]:
# 1. Preprocess the reviews_df dataframe to prepare it for the following questions.
???

In [1]:
import string
import gzip
from collections import Counter

import numpy as np
import pandas as pd

import matplotlib
import matplotlib.pyplot as plt
import matplotlib. cm as cm

import nltk
from nltk. sentiment import SentimentIntensityAnalyzer

import watermark
%matplotlib inline
%load_ext watermark

In [2]:
# used for natural language processing {NLTK: Natural Language Tool-Kit} 
from nltk.stem import WordNetLemmatizer, SnowballStemmer, LancasterStemmer
from nltk.stem.porter import *
import nltk
nltk.download('wordnet')

# Set a seed to reproduce the results later. Seed used here is '2019'
import numpy as np

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/aishatolatunji/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [3]:
import warnings
warnings.filterwarnings('ignore')

import gzip
from collections import Counter
from pprint import pprint

import pandas as pd
import numpy as np
np.random.seed(123456)

import matplotlib
import matplotlib.pyplot as plt

from tqdm import tqdm

import keras
import tensorflow as tf
from keras.datasets import imdb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Flatten
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D
#from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence

import watermark
%load_ext watermark
%matplotlib inline

The watermark extension is already loaded. To reload it, use:
  %reload_ext watermark


In [4]:
import warnings
warnings.filterwarnings('ignore')

import gzip
from collections import Counter
from pprint import pprint

import pandas as pd
import numpy as np
np.random.seed(123456)

import matplotlib
import matplotlib.pyplot as plt

from tqdm import tqdm



In [5]:
import warnings
warnings.filterwarnings('ignore')

import gzip
from collections import Counter
from pprint import pprint

import pandas as pd
import numpy as np
np.random.seed(42)

import matplotlib
import matplotlib.pyplot as plt

from tqdm import tqdm
tqdm.pandas()

np.random.seed(123)

import keras
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN
from keras.layers import LSTM, GRU
from keras.layers import Dense
from keras.datasets import imdb
from keras.preprocessing import sequence

import watermark

%load_ext watermark
%matplotlib inline

The watermark extension is already loaded. To reload it, use:
  %reload_ext watermark


In [6]:
plt.style.use('seaborn')

In [7]:
Reviews_df = pd.read_csv("AmazonReviews.csv")

In [8]:
Reviews_df.head()

Id   ProductId          UserId                      ProfileName  \
0   1  B001E4KFG0  A3SGXH7AUHU8GW                       delmartian   
1   2  B00813GRG4  A1D87F6ZCVE5NK                           dll pa   
2   3  B000LQOCH0   ABXLMWJIXXAIN  Natalia Corres "Natalia Corres"   
3   4  B000UA0QIQ  A395BORC6FGVXV                             Karl   
4   5  B006K2ZZ7K  A1UQRSCLF8GW1T    Michael D. Bigham "M. Wassir"   

   HelpfulnessNumerator  HelpfulnessDenominator  Score        Time  \
0                     1                       1      5  1303862400   
1                     0                       0      1  1346976000   
2                     1                       1      4  1219017600   
3                     3                       3      2  1307923200   
4                     0                       0      5  1350777600   

                 Summary                                               Text  
0  Good Quality Dog Food  I have bought several of the Vitality canned d...  
1      Not as Advertised  Product arrived labeled as Jumbo Salted Peanut...  
2  "Delight" says it all  This is a confection that has been around a fe...  
3         Cough Medicine  If you are looking for the secret ingredient i...  
4            Great taffy  Great taffy at a great price.  There was a wid...

In [9]:
len(Reviews_df)

25000

In [10]:
Reviews_df.loc[Reviews_df['Score'] >= 3, 'sentiment'] = 1
Reviews_df.loc[Reviews_df['Score'] < 3, 'sentiment'] = 0 

In [ ]:
# Data Cleaning

In [11]:
Reviews_df['Text'] = Reviews_df['Text'].str.lower()

In [12]:
Reviews_df['Text'] = Reviews_df['Text'].str.replace(r'[][]+', '', regex=True)

In [13]:
Reviews_df['Text'] = Reviews_df['Text'].str.replace('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', '')



In [14]:
#Reviews_df['Text'] = Reviews_df['Text'].str.replace('[^\w\s]','')

In [14]:
Reviews_df['Text']=Reviews_df['Text'].str.replace('[{}]'.format(string.punctuation), '')

In [15]:
Reviews_df['Text'] = Reviews_df['Text'].str.replace('\d+', '')



In [ ]:
# procedures

In [16]:
from keras_pos_embd import PositionEmbedding

In [17]:
from keras_pos_embd import TrigPosEmbedding



In [18]:
from sklearn.model_selection import train_test_split

In [19]:
Y = Reviews_df['sentiment']

In [20]:
Y[2]

1.0

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer
# The script below uses tfidf and then divides our data into 80% for the training set and 20% for the testing set
vectorizer = TfidfVectorizer(stop_words='english', max_features=500, ngram_range=(1,1))
X_train = vectorizer.fit_transform(Reviews_df['Text'])[:20000].A 
Y_train = Reviews_df.sentiment[:20000] 
X_test = vectorizer.fit_transform(Reviews_df['Text'])[20000:].A
Y_test = Reviews_df.sentiment[20000:]
# The train set will be used to train our learning model while the testing set will be used to evalute hoe well our model performs

In [22]:
max_words = 500
X_train = keras.utils.pad_sequences(X_train, maxlen=max_words)
X_test = keras.utils.pad_sequences(X_test, maxlen=max_words)

In [ ]:
# 2. Build a SimpleRNN Network to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy
???

In [23]:
model = Sequential()
#model.add(keras.layers.Embedding(10000, 32, input_length=max_words))
model.add(Embedding(10000, 32, input_length=max_words))
model.add(SimpleRNN(32))
model.add(Dense(1, activation='sigmoid'))
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 500, 32)           320000    
                                                                 
 simple_rnn (SimpleRNN)      (None, 32)                2080      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 322,113
Trainable params: 322,113
Non-trainable params: 0
_________________________________________________________________


2022-11-13 00:08:00.373792: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [24]:
model.compile(loss='binary_crossentropy',
             optimizer='adam',
             metrics=['accuracy'])

In [25]:
history = model.fit(X_train, Y_train,
                   validation_data=(X_test, Y_test),
                   epochs=10,
                   batch_size=128,
                   verbose=1)

Epoch 1/10
157/157 [==============================] - 9s 55ms/step - loss: 0.4501 - accuracy: 0.8486 - val_loss: 0.4259 - val_accuracy: 0.8486
Epoch 2/10
157/157 [==============================] - 8s 54ms/step - loss: 0.4180 - accuracy: 0.8528 - val_loss: 0.4256 - val_accuracy: 0.8486
Epoch 3/10
157/157 [==============================] - 8s 54ms/step - loss: 0.4180 - accuracy: 0.8528 - val_loss: 0.4251 - val_accuracy: 0.8486
Epoch 4/10
157/157 [==============================] - 8s 54ms/step - loss: 0.4180 - accuracy: 0.8528 - val_loss: 0.4253 - val_accuracy: 0.8486
Epoch 5/10
157/157 [==============================] - 9s 55ms/step - loss: 0.4180 - accuracy: 0.8528 - val_loss: 0.4252 - val_accuracy: 0.8486
Epoch 6/10
157/157 [==============================] - 8s 53ms/step - loss: 0.4181 - accuracy: 0.8528 - val_loss: 0.4251 - val_accuracy: 0.8486
Epoch 7/10
157/157 [==============================] - 8s 53ms/step - loss: 0.4179 - accuracy: 0.8528 - val_loss: 0.4252 - val_accuracy: 0.8486

In [ ]:
# 3. Build a Gated Recurrent Unit to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy
???

In [26]:
model = Sequential()
model.add(Embedding(10000, 32, input_length=max_words))
model.add(GRU(32))
model.add(Dense(1, activation='sigmoid'))
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_1 (Embedding)     (None, 500, 32)           320000    
                                                                 
 gru (GRU)                   (None, 32)                6336      
                                                                 
 dense_1 (Dense)             (None, 1)                 33        
                                                                 
Total params: 326,369
Trainable params: 326,369
Non-trainable params: 0
_________________________________________________________________


In [27]:
model.compile(loss='binary_crossentropy',
             optimizer='adam',
             metrics=['accuracy'])

In [28]:
history2 = model.fit(X_train, Y_train,
                   validation_data=(X_test, Y_test),
                   epochs=10,
                   batch_size=128,
                   verbose=1)

Epoch 1/10
157/157 [==============================] - 20s 123ms/step - loss: 0.4437 - accuracy: 0.8528 - val_loss: 0.4269 - val_accuracy: 0.8486
Epoch 2/10
157/157 [==============================] - 19s 120ms/step - loss: 0.4184 - accuracy: 0.8528 - val_loss: 0.4252 - val_accuracy: 0.8486
Epoch 3/10
157/157 [==============================] - 19s 123ms/step - loss: 0.4184 - accuracy: 0.8528 - val_loss: 0.4257 - val_accuracy: 0.8486
Epoch 4/10
157/157 [==============================] - 19s 122ms/step - loss: 0.4181 - accuracy: 0.8528 - val_loss: 0.4251 - val_accuracy: 0.8486
Epoch 5/10
157/157 [==============================] - 19s 121ms/step - loss: 0.4183 - accuracy: 0.8528 - val_loss: 0.4252 - val_accuracy: 0.8486
Epoch 6/10
157/157 [==============================] - 20s 124ms/step - loss: 0.4186 - accuracy: 0.8528 - val_loss: 0.4278 - val_accuracy: 0.8486
Epoch 7/10
157/157 [==============================] - 19s 123ms/step - loss: 0.4182 - accuracy: 0.8528 - val_loss: 0.4252 - val_ac

In [ ]:
# 4. Build a Long Short-Term Memory Model to predict the sentiment of each review. Then, evaluate the performance of the model in terms of loss and accuracy
???

In [29]:
model = Sequential()
model.add(Embedding(10000, 32, input_length=max_words))
model.add(LSTM(32))
model.add(Dense(1, activation='sigmoid'))
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_2 (Embedding)     (None, 500, 32)           320000    
                                                                 
 lstm (LSTM)                 (None, 32)                8320      
                                                                 
 dense_2 (Dense)             (None, 1)                 33        
                                                                 
Total params: 328,353
Trainable params: 328,353
Non-trainable params: 0
_________________________________________________________________


In [30]:
model.compile(loss='binary_crossentropy',
             optimizer='adam',
             metrics=['accuracy'])

In [31]:
history3 = model.fit(X_train, Y_train,
                   validation_data=(X_test, Y_test),
                   epochs=10,
                   batch_size=128,
                   verbose=1)

Epoch 1/10
157/157 [==============================] - 21s 131ms/step - loss: 0.4448 - accuracy: 0.8528 - val_loss: 0.4256 - val_accuracy: 0.8486
Epoch 2/10
157/157 [==============================] - 20s 130ms/step - loss: 0.4183 - accuracy: 0.8528 - val_loss: 0.4252 - val_accuracy: 0.8486
Epoch 3/10
157/157 [==============================] - 20s 130ms/step - loss: 0.4181 - accuracy: 0.8528 - val_loss: 0.4253 - val_accuracy: 0.8486
Epoch 4/10
157/157 [==============================] - 20s 129ms/step - loss: 0.4183 - accuracy: 0.8528 - val_loss: 0.4263 - val_accuracy: 0.8486
Epoch 5/10
157/157 [==============================] - 20s 130ms/step - loss: 0.4183 - accuracy: 0.8528 - val_loss: 0.4264 - val_accuracy: 0.8486
Epoch 6/10
157/157 [==============================] - 20s 129ms/step - loss: 0.4181 - accuracy: 0.8528 - val_loss: 0.4256 - val_accuracy: 0.8486
Epoch 7/10
157/157 [==============================] - 20s 129ms/step - loss: 0.4183 - accuracy: 0.8528 - val_loss: 0.4254 - val_ac

i. SimpleRnn network at the 10th iteration(i.e it used 10 epochs) produced an accuracy of 0.8528 and Val_accuracy of 0.8486

ii. SimpleRnn network at the 10th iteration(i.e it used 10 epochs)  produced a loss  of 0.4182 and Val_loss of 0.4252



i. Gated Recurrent unit at the 10th iteration(i.e it used 10 epochs) produced an accuracy of 0.8528 and Val_accuracy of 0.8486

ii. Gated Recurrent unit at the 10th iteration(i.e it used 10 epochs)  produced a loss  of 0.4182 and Val_loss of 0.4254




i. longshort term memmory model at the 10th iteration(i.e it used 10 epochs) produced an accuracy of 0.8528 and Val_accuracy of 0.8486

ii. longshort term memmory model at the 10th iteration(i.e it used 10 epochs)  produced a loss  of 0.4183 and Val_loss of 0.4251



In [32]:
# 5. Which model performs better and why?

The three models seems to provide the same accuracy and val accuracy but the longshort term memmory model seems to be the best model here because at the 10th iteration, it seems to minimize the val_loss at the rate of 0.4251 better the remaining models, so i would conclude that the long short term memmory model performs better.

<div class="alert alert-block alert-info">

# Part 2: Reflection
    
</div>

As a second step—after answering the questions, include the following:
1. A reflection of your experience performing the activity. 
2. A reflection on the importance of learning this activity.
**Note:** include your reflection in this notebook as markdown cells. 

A reflection of your experience performing the activity.

Working on this particular project drove me back to machine learning on how to split data set using validation approach and it's important to avoid over fitting.

i. i also understood that the two losses(both loss and val_loss) are decreasing and the accuracy (acc and val_acc) are increasing which indicates the modelling is trained in a good way.

ii. The val_acc is the measure of how good the predictions of your model are. in this case the project used 10 epochs i.e it iterated 10 times, this study also explains to me we do not need to increase the iteration times as it would still produce the same result.
iv. i also noticed that the accuaracy and the val accuracy remained static which means it's good at 10th iteration

A reflection on the importance of learning this activity.

In deep learning, the loss is the value that a neural network is trying to minimize: it's the distance between the ground truth and the predictions. In order to minimize this distance, the neural network learns by adjusting weights and biases in a manner that reduces the loss.

In classification, it's a little more complicated, but very similar. Predicted classes are based on probability. The loss is therefore also based on probability. In classification, the neural network minimizes the likelihood to assign a low probability to the actual class. The loss is typically categorical_crossentropy

<div class="alert alert-block alert-info">

# Submission
    
</div>

Submit **Lab8.ipynb** to the **Lab 8 - Sequence Modeling (Dropbox)** on D2L by the due date. 

<div class="alert alert-block alert-info">

# Grading Rubric
    
</div>

|Criterion	|Excellent	|Good	|Average	|Below Average	|Poor	|No Attempt|
|:--	|:--	|:--	|:--	|:-- 	|:--	|:-- |
|**Part 1:** Activity-Question 1|	**10 points**- Completes all aspects of the question correctly	|**8 points**- Completes most aspects of the question correctly |**6 points**- Completes aspects of the question  correctly, and some incorrectly	|**4 points**- Completes most aspects of the question  incorrectly or does not attempt many aspects|	**2 points**- Minimal effort or completes a few aspects of the question or very few correctly|**0 points**- Did not complete the question|
|**Part 1:** Activity-Question 2|	**10 points**- Completes all aspects of the question correctly	|**8 points**- Completes most aspects of the question correctly |**6 points**- Completes aspects of the question  correctly, and some incorrectly	|**4 points**- Completes most aspects of the question  incorrectly or does not attempt many aspects|	**2 points**- Minimal effort or completes a few aspects of the question or very few correctly|**0 points**- Did not complete the question|
|**Part 1:** Activity-Question 3|	**10 points**- Completes all aspects of the question correctly	|**8 points**- Completes most aspects of the question correctly |**6 points**- Completes aspects of the question  correctly, and some incorrectly	|**4 points**- Completes most aspects of the question  incorrectly or does not attempt many aspects|	**2 points**- Minimal effort or completes a few aspects of the question or very few correctly|**0 points**- Did not complete the question|
|**Part 1:** Activity-Question 4|	**10 points**- Completes all aspects of the question correctly	|**8 points**- Completes most aspects of the question correctly |**6 points**- Completes aspects of the question  correctly, and some incorrectly	|**4 points**- Completes most aspects of the question  incorrectly or does not attempt many aspects|	**2 points**- Minimal effort or completes a few aspects of the question or very few correctly|**0 points**- Did not complete the question|
|**Part 1:** Activity-Question 5|	**10 points**- Completes all aspects of the question correctly	|**8 points**- Completes most aspects of the question correctly |**6 points**- Completes aspects of the question  correctly, and some incorrectly	|**4 points**- Completes most aspects of the question  incorrectly or does not attempt many aspects|	**2 points**- Minimal effort or completes a few aspects of the question or very few correctly|**0 points**- Did not complete the question|
|**Part 2:** Reflection|**10 points**- Reflection clearly ties to the module content; experience and importance clearly laid out|**8 points**- Reflection mostly ties to the module content; experience & importance are discussed|**6 points**- Reflection ties minimally to the module content; experience & importance are discussed but not thoroughly|**4 points**- Reflection does not tie to the module content; experience & importance are minimally discussed|**2 points**- Minimal effort to tie to content; minimal effort to describe experience/ importance|**0 points**- Did not complete the reflection